File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/template_method/solution/DpStack.m
% DpStack.m

classdef DpStack < handle
	properties (GetAccess = private, SetAccess = private)
		Link % DpLink
	end

	methods
		function result = isEmpty(theDpStack)
			result = isempty(theDpStack.Link);
		end

		function push(theDpStack, aValue)
			theDpStack.Link = DpLink(aValue, theDpStack.Link);
		end

		function pop(theDpStack)
			if isEmpty(theDpStack)
				error('Cannot pop an empty DpStack.');
			end

			theDpStack.Link = theDpStack.Link.Next;
		end

		function result = top(theDpStack)
			if isEmpty(theDpStack)
				error('Cannot top an empty DpStack.');
			end

			result = theDpStack.Link.Value;
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/data_structures/dp_stack.java
// dp_stack.java

public class dp_stack<T>
{
	public dp_stack()
	{
		my_link = null;
	}

	public boolean is_empty()
	{
		return (my_link == null);
	}

	public void push(T a_value)
	{
		my_link = new dp_link<T>(a_value, my_link);
	}

	public void pop()
	{
		if (is_empty())
		{
			throw new dp_exception("dp_stack<T>.pop() - The stack is empty.");
		}
		my_link = my_link.get_next();
	}

	public T top()
	{
		if (is_empty())
		{
			throw new dp_exception("dp_stack<T>.top() - The stack is empty.");
		}
		return my_link.get_value();
	}

	private dp_link<T> my_link;
}
